src/sourcetransform.rs

```diff
@@ -235,44 +242,236 @@ impl<T> proto::source_transform_server::SourceTransform for SourceTransformerSer
+) -> bool
+where
+    T: SourceTransformer + Send + Sync + 'static,
+{
+    let transform_request = match transform_request {
+        Ok(None) => return false,
+        Ok(Some(val)) => val,
+        Err(val) => {
+            warn!("Received gRPC error from sender: {val:?}");
+            return false;
+        }
+    };
+    tokio::spawn(run_transform(
+        handler,
+        transform_request,
+        stream_response_tx,
+        error_tx,
+        token,
+    ));
+    true
+}
+
+// Calls the user implemented transform function on the request.
+async fn run_transform<T>(
+    handler: Arc<T>,
+    transform_request: proto::SourceTransformRequest,
+    stream_response_tx: mpsc::Sender<Result<SourceTransformResponse, Status>>,
+    error_tx: mpsc::Sender<Error>,
+    token: CancellationToken,
+) where
+    T: SourceTransformer + Send + Sync + 'static,
+{
+    let Some(request) = transform_request.request else {
+        error_tx
+            .send(SourceTransformerError(ErrorKind::InternalError(
+                "Transform request can not be none".to_string(),
+            )))
+            .await
+            .expect("Sending error on channel");
+        return;
+    };
+
+    let message_id = request.id.clone();
+
+    // A new task is spawned to catch the panic
+    let udf_tranform_task = tokio::spawn({
+        let handler = handler.clone();
+        let token = token.child_token();
+        async move {
+            tokio::select! {
+                _ = token.cancelled() => None,
+                messages = handler.transform(request.into()) => Some(messages),
+            }
+        }
+    });
+
+    let messages = match udf_tranform_task.await {
+        Ok(messages) => messages,
+        Err(e) => {
+            tracing::error!("Failed to run transform function: {e:?}");
+            error_tx
+                .send(SourceTransformerError(ErrorKind::UserDefinedError(
+                    "panic in transform UDF".to_string(),
+                )))
+                .await
+                .expect("Sending error on channel");
+            return;
+        }
+    };
+
+    let Some(messages) = messages else {
+        // CancellationToken is cancelled
+        return;
+    };
+
+    let send_response_fut = stream_response_tx.send(Ok(SourceTransformResponse {
+        results: messages.into_iter().map(|msg| msg.into()).collect(),
+        id: message_id,
+        handshake: None,
+    }));
+
+    let stream_response_result = tokio::select! {
+        result = send_response_fut => result,
+        _ = token.cancelled() => return,
+    };
+
+    let Err(e) = stream_response_result else {
+        return;
+    };
+
+    let send_err_fut = error_tx.send(SourceTransformerError(ErrorKind::InternalError(format!(
+        "sending source transform response over gRPC stream: {e:?}"
+    ))));
+
+    tokio::select! {
+        err = send_err_fut => err.expect("Sending error on channel"),
+        _ = token.cancelled() => (),
+    }
```

Comment:
Do we really need to handle cancellation token while writing to channels?

Reply:
Removed